staging/src/k8s.io/externaljwt/pkg/plugin/plugin.go

```diff
@@ -0,0 +1,263 @@
+	ejspb "k8s.io/externaljwt/apis/v1alpha1"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+type VerificationKeys struct {
+	Keys          []serviceaccount.PublicKey
+	DataTimestamp time.Time
+}
+
+type Metadata struct {
+	MaxTokenExpirationSeconds int
+}
+
+// New sets up all dependencies for the plugin and cache, in
+// order to minimize changes to upstream files.
+func New(ctx context.Context, issuer, socketPath string) (*Plugin, *KeyCache, error) {
+	conn, err := grpc.Dial(
+		socketPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(func(ctx context.Context, path string) (net.Conn, error) {
+			return (&net.Dialer{}).DialContext(ctx, "unix", path)
+		}),
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("while dialing go/gke-external-token-signing grpc socket: %w", err)
+	}
+
+	plugin := newPlugin(issuer, conn)
+
+	initialFillCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
+
+	if err := plugin.keyCache.InitialFill(initialFillCtx); err != nil {
+		return nil, nil, fmt.Errorf("while initially filling go/gke-external-token-signing key cache: %w", err)
+	}
+	go plugin.keyCache.StartPeriodicSync(ctx)
+
+	go func() {
+		<-ctx.Done()
+		conn.Close()
+	}()
+
+	return plugin, plugin.keyCache, nil
+}
+
+// enables pluging in an external jwt signer.
+type Plugin struct {
+	iss      string
+	client   ejspb.ExternalJWTSignerClient
+	keyCache *KeyCache
+}
+
+// newPlugin constructs an implementation of external JWT signer plugin.
+func newPlugin(iss string, conn *grpc.ClientConn) *Plugin {
+	client := ejspb.NewExternalJWTSignerClient(conn)
+	plugin := &Plugin{
+		iss:    iss,
+		client: client,
+	}
+	plugin.keyCache = newKeyCache(plugin)
+	return plugin
+}
+
+// GenerateToken creates a service account token with the provided claims by
+// calling out to the external signer binary.
+func (p *Plugin) GenerateToken(claims *jwt.Claims, privateClaims interface{}) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	claims.Issuer = p.iss
+
+	mergedClaims, err := mergeClaims(claims, privateClaims)
+	if err != nil {
+		return "", fmt.Errorf("while merging claims: %w", err)
+	}
+
+	payload := base64.RawURLEncoding.EncodeToString(mergedClaims)
+
+	request := &ejspb.SignJWTRequest{
+		Claims: payload,
+	}
+
+	response, err := p.client.Sign(ctx, request)
+	if err != nil {
+		return "", fmt.Errorf("while signing jwt: %w", err)
+	}
+
+	if err := p.validateJWTHeader(response); err != nil {
+		return "", fmt.Errorf("while validating header: %w", err)
+	}
+
+	return response.GetHeader() + "." + payload + "." + response.GetSignature(), nil
+}
+
+// GetTokenVerificationKeys returns a map of supported external keyIDs to keys
+// the keys are PKIX-serialized.
+func (p *Plugin) GetTokenVerificationKeys(ctx context.Context) (*VerificationKeys, int, error) {
+	ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
```

Comment:
I mean, it will get hardcoded somewhere on k8s. The only source for a dynamic value that makes sense is (I think) the external signer. We can maybe do that in the next iteration?

Reply:
Doing it per offline discussion.